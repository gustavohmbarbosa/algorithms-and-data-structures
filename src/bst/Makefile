all:
	make test f=1 && make test f=2 && make test f=3 && make test f=4 && make test f=5 && make test f=7 && make test f=8 && make test f=9 && make test f=10 && make test f=11 && make test f=13

main: main.o bst.o
	gcc -o main.bin main.o bst.o

test: file
	diff f$(f).test.out ./tests/f$(f).out

file: main
	./main.bin < ./tests/f$(f).in -lh > f$(f).test.out 

main.o: main.c
	gcc -c main.c

bst.o: bst.c
	gcc -c $<

clean:
	rm -f *.o *.out *.bin