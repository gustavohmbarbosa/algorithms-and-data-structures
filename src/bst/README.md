# How to run the program
```shell
make main && ./main.bin
```

# How to run the tests
```shell
# run all tests
make

# run a specif test
make test f=number_of_test

# create a output file test, without verify if matches
make file f=number_of_test
```