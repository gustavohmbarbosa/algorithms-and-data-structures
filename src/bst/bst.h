#ifndef BST_H
#define BST_H

typedef struct Node {
    int value;
    struct Node *right;
    struct Node *left;
} Node;

typedef Node *Tree;

void insert(Tree *root, int value);

void print_pre_order(Tree root);

void print_in_order(Tree root);

void print_post_order(Tree root);

void print_reverse_order(Tree root);

int even_amount(Tree root);

#endif