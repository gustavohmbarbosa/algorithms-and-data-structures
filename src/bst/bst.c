#include <stdio.h>
#include <stdlib.h>
#include "bst.h"

void insert(Tree *root, int value) {
    if (*root == NULL) {
        *root = (Tree) malloc(sizeof(Node)); 
        (*root)->right = NULL;
        (*root)->left = NULL;
        (*root)->value = value;
        return;
    }

    if (value > (*root)->value) {
        return insert(&(*root)->right, value);
    }

    return insert(&(*root)->left, value);
}


void _print_pre_order(Tree root) {
    if (root == NULL) {
        return;
    }

    printf("[%d]", root->value);
    _print_pre_order(root->left);
    _print_pre_order(root->right);
}

void print_pre_order(Tree root) {
    _print_pre_order(root);
    printf("\n");
}


void _print_in_order(Tree root) {
    if (root == NULL) {
        return;
    }

    _print_in_order(root->left);
    printf("[%d]", root->value);
    _print_in_order(root->right);
}

void print_in_order(Tree root) {
    _print_in_order(root);
    printf("\n");
}

void _print_post_order(Tree root) {
    if (root == NULL) {
        return;
    }

    _print_post_order(root->left);
    _print_post_order(root->right);
    printf("[%d]", root->value);
}

void print_post_order(Tree root) {
    _print_post_order(root);
    printf("\n");
}

void _print_reverse_order(Tree root) {
    if (root == NULL) {
        return;
    }

    _print_reverse_order(root->right);
    printf("[%d]", root->value);
    _print_reverse_order(root->left);
}

void print_reverse_order(Tree root) {
    _print_reverse_order(root);
    printf("\n");
}

int even_amount(Tree root) {
    if (root == NULL) {
        return 0;
    }

    int result = even_amount(root->left) + even_amount(root->right);
    if (root->value % 2 == 0) {
        result++;
    }

    return result;
}

int predecessor(Tree root, int value) {
    int result = value;
    
    Tree temp = root;
    int has_value = 0;
    while (temp != NULL) {
        if (temp->value < value) {
            if (temp->value > result || result == value) {
                result = temp->value;
            }

            temp = temp->right;
            continue;
        }

        if (temp->value > value) {
            temp = temp->left;
            continue;
        }

        if (temp->value == value) {
            has_value = 1;
            temp = temp->left;
            continue;
        }

        break;
    }

    return (has_value && result != value) ? result : -1;
}


int find_parent(Tree root, int value, int parent) {
    if (root == NULL) {
        return -1;
    }

    if (root->value > value) {
        return find_parent(root->left, value, root->value);
    }

    if (root->value < value) {
        return find_parent(root->right, value, root->value);
    }

    return parent;
}

int parent(Tree root, int value) {
    if (root == NULL || root->value == value) {
        return -1;
    }
 
    return find_parent(root, value, value);
}

int exists(Tree root, int value) {
    if (root == NULL) {
        return 0;
    }

    if (root->value == value) {
        return 1;
    }

    if (value > root->value) {
        return exists(root->right, value);
    }

    if (value < root->value) {
        return exists(root->left, value);
    }

    return 0;
}
