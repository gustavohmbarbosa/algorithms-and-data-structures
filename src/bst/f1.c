#include <stdlib.h>
#include <stdio.h>
#include "bst.h"

int main(void *args) {
    Tree tree = NULL;
    insert(&tree, 40);
    insert(&tree, 20);
    insert(&tree, 50);

    print_pre_order(tree);
    printf("\n");
    return 0;
}
