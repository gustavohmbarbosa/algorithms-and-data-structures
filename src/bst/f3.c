#include <stdlib.h>
#include <stdio.h>
#include "bst.h"

int main(void *args) {
    Tree tree = NULL;
    insert(&tree, 23);
    insert(&tree, 80);
    insert(&tree, 21);
    insert(&tree, 24);
    insert(&tree, 214);
    insert(&tree, 14);
    insert(&tree, 11);

    print_in_order(tree);
    printf("\n");
    return 0;
}
