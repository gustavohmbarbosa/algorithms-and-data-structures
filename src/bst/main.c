#include<stdio.h>
#include<stdlib.h>
#include "bst.h"

int main (void *args) {
    int option, value;
    Tree tree = NULL;
    while (1) {
        scanf("%d", &option);
        switch (option) {
            case 1:
                scanf("%d", &value);
                tree_insert(&tree, value);
                break;
            case 2:
                tree_print_pre_order(tree);
                break;
            case 3:
                tree_print_in_order(tree);
                break;
            case 4:
                tree_print_post_order(tree);
                break;
            case 5:
                tree_print_reverse_order(tree);
                break;
            case 6:
                printf("%d\n", tree_even_amount(tree));
                break;
            case 7:
                scanf("%d", &value);
                printf("%d\n", tree_predecessor(tree, value));
                break;
            case 8:
                scanf("%d", &value);
                printf("%d\n", tree_parent(tree, value));
                break;
            case 9:
                scanf("%d", &value);
                tree_remove(&tree, value);
                break;
            case 10:
                printf("%d\n", tree_even_sum(tree));
                break;
            case 11:
                scanf("%d", &value);
                tree_prune(&tree, value);
                break;
            case 12:
                tree_pow_sheets(tree);
                break;
            case 13:
                scanf("%d", &value);
                printf("%d\n", tree_exists(tree, value));
                break;
            case 14:
                scanf("%d", &value);
                tree_print_children_in_order(tree, value);
                break;
            case 99:
                exit(0);
            default:
                break;
        }
    }

    return 0;
}