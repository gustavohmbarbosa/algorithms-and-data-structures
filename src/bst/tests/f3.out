[11][14][21][23][24][80][214]
